### PR TITLE
Reduce dask task count, refactor regrid dispatch

### LINKED ISSUE #117: FutureWarning: ``output_sizes`` should be given in the ``dask_gufunc_kwargs`` parameter
FYI came across this in some code I was using that uses xESMF under the hood

```
/opt/userenvs/ray.bell/main/lib/python3.9/site-packages/xesmf/frontend.py:522: FutureWarning: ``output_sizes`` should be given in the ``dask_gufunc_kwargs`` parameter. It will be removed as direct parameter in a future version.
  dr_out = xr.apply_ufunc(
```

https://github.com/pangeo-data/xESMF/blob/master/xesmf/frontend.py#L522

### LINKED ISSUE #127: Very large Dask graph size with high resolution grids
Regridding from 0.1 to 0.25 degree global grids is creating a very large task graph with xesmf 0.6.1, causing problems when evaluating the result.

Sample code:
```python
import xarray
import xesmf
import dask.array

a = xesmf.util.grid_global(0.1,0.1)
b = xesmf.util.grid_global(0.25,0.25)
r = xesmf.Regridder(a,b,method='bilinear')

data = dask.array.random.random((720,a.dims['y'], a.dims['x']), chunks=(12,-1,-1))
da = xarray.DataArray(data, dims=['time','y','x'], coords={'y': a.y, 'x': a.x})

r(da).data
```

Evaluated versions at https://nbviewer.org/gist/ScottWales/4fe0e9a5725b5a3bf07ab94e80778846

With 0.6.0, the regridded data has 240 Dask tasks. With 0.6.1 the number has exploded to 409,841 tasks, and appears to scale with the grid resolution - low resolution regrid task graphs have a reasonable size.

Expected behaviour is that the Dask task graph created by xesmf is a reasonably small multiple of the number of chunks in the input data, and that it is independent of grid resolution. Having a too large task graph slows down Dask as it prepares the tasks to run and is causing excess memory use when the result is evaluated.

### PR DESCRIPTION
**Pull request by aulemahal**
This fixes #127 and fixes #117 (maybe that one was already fixed?).

At the beginning, I tried to explore @dcherian's idea of using dask's `tensordot` directly instead of wrapping everything in a  `xr.apply_ufunc` with `dask='parallelized'`. Sadly, that didn't go as well as planned : the performance was not better, sometimes it was equal, sometimes worst. But, building on that, I moved things around to see if it could improve.  I'm a bit disappointed as this final solution is not _that_ much more performant than before...

The dask graph issue (127) is fixed explicitly by wrapping the weights into a dask array ourselves. I could have done just that, but the changes here have the following bonuses:
- Reorder the checks and the computation : shape checks are now done _before_ the computation, which means a wrongly shaped input using dask will raise an error on the line calling the regridder, instead of further down in the `load()` or `to_netcdf` call.
- The number of dasks added by the Regridder is now Nchunks + 1, no more. It's slightly better than before (master).
- It erases the fix of #122. Dask is now made aware of the output dtype outside of xarray, so dtype is preseved on dataset without additionnal code. 

Because of this rewrite and constraints given by `regrid_dataset`, it was easier to have a single `regrid_array` that handles the switch between dask and numpy arrays. That rendered `regrid_numpy` and `regrid_dask` kinda useless, so I took the decision to deprecated them. We can remove the warnings if you think this breaking change is not necessary.

## Performances
As said above, the overall performance of this branch is quite similar to the one of the master IF we set the dask config to avoid the bug of #127. In general, it reduces the number of task in the tree, but that improvement was diluted in my tests because dask is able to optimize said tree by merging tasks. Memory-wise, the two branches are quite equivalent, both are better than v0.6.0. 

## Going forward
Overall this is a lot of lines changed for a bug fix, but it opens possibilities for handling of spatial chunks. Without `dask='parallelized`, it could be possible to have a chunked output for very large grids?

I observed serious memory usage when `skipna=True`. I'm pretty sure this could be improved by merging some operations together?

Also, now that the weights are in a sparse DataArray, we could make them 4-D instead of 2D, maybe that could reduce the  `reshape` calls and improve performance?

All things to do in further PRs.

## TL;DR
Thisfixes a bug, but introduces many less important changes because it was coded incrementally.  If requested, I can redo another PR with only the bug-fixing change.

## EDIT:
Other changes:
- The `assert` statements in prior to the regridding have been changed into ValueErrors, for more pythonic code.

### CODE REVIEW

File: xesmf/smm.py [hunk 1]
```diff
@@ -101,6 +101,29 @@ def _parse_coords_and_values(indata, n_in, n_out):
     return xr.DataArray(sps.COO(crds, s, (n_out, n_in)), dims=('out_dim', 'in_dim'), name='weights')
 
 
+def check_shapes(arr_shape, w_shape, shape_in, shape_out):
+
```
Comment (huard) [inline xesmf/smm.py:105]: Docstring

File: xesmf/smm.py [hunk 2]
```diff
@@ -101,6 +101,29 @@ def _parse_coords_and_values(indata, n_in, n_out):
...
+    assert shape_horiz == shape_in, (
+        'The horizontal shape of input data is {}, different from that of'
+        'the regridder {}!'.format(arr_shape[-2:], shape_in)
+    )
+
+    assert shape_in[0] * shape_in[1] == w_shape[1], 'ny_in * nx_in should equal to weights.shape[1]'
+
+    assert (
+        shape_out[0] * shape_out[1] == w_shape[0]
+    ), 'ny_out * nx_out should equal to weights.shape[0]'
```
Comment (huard) [inline xesmf/smm.py:124]: replace with try/except

> Second, Python interpreter users can disable assertion statements. Therefore, a number of problems can arise if you rely on assertions too heavily for error handling and a user disables assertions.

**Review by huard — CHANGES_REQUESTED** [review]
Minor changes.

**Comment by huard** [comment]
@ScottWales Could you check this out and confirm that it works for you ?

**Comment by dcherian** [comment]
>  Sadly, that didn't go as well as planned : the performance was not better, sometimes it was equal, sometimes worst.

Well the other way of looking at it is that you've enabled a major new feature (chunked spatial dimensions) without much of a performance regression 😉 

Maybe it's possible to improve tensordot for the case where there is only one chunk along spatial dimensions?

**Comment by aulemahal** [comment]
AFAIK, the performance issues where not from `tensordot` itself. In my first test, I was doing everything a "duck type" manner, i.e. : all operations on `indata` in `_regrid_array` and `apply_weights` were chosen to work with numpy and dask arrays alike. This way, dask's tensordot was used, but it added many `reshape`, `transpose` and a `astype` task. I think it is this high number of very small tasks that cancelled the potential performance gain.

**Review by huard — APPROVED** [review]

**Review by raphaeldussin — APPROVED** [review]

**Comment by raphaeldussin** [comment]
great job @aulemahal
